File: OpenCVによる画像処理入門/第3版/第8刷/sampleprogram/8-5-4.py
import cv2
import math
import numpy as np

file_src = 'src_bin_circle.png'
file_dst = 'dst.png'

img_src = cv2.imread(file_src, cv2.IMREAD_GRAYSCALE)

cv2.namedWindow('src')
# cv2.namedWindow('dst')

contours = cv2.findContours(img_src, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)[0]
area = cv2.contourArea(contours[0])  # 面積
perimeter = cv2.arcLength(np.array(contours[0]), True)  # 周囲長
roundness = 4 * np.pi * area / perimeter / perimeter  # 円形度
print(area, perimeter, roundness)

cv2.imshow('src', img_src)  # 入力画像を表示
# cv2.imshow('dst', img_dst) # 出力画像を表示
# cv2.imwrite(file_dst, img_dst) # 処理結果の保存

cv2.waitKey(0)  # キー入力待ち
cv2.destroyAllWindows()

File: OpenCVによる画像処理入門/第3版/第8刷/sampleprogram/8-5-4.cpp
#define _USE_MATH_DEFINES
#include <iostream>
#include <cmath>
#include <opencv2/opencv.hpp>
using namespace std;
using namespace cv;
string win_src = "src";
string win_dst = "dst";

int main()
{
  string file_src = "src_bin_circle.png"; // 入力画像のファイル名
  string file_dst = "dst.png"; // 出力画像のファイル名
  Mat img_src = imread(file_src, IMREAD_GRAYSCALE); // 入力画像（グレースケール）の読み込み

  Mat img_dst;
  if(!img_src.data) {
    cout << "error" << endl;
    return -1;
  }

  // ウインドウ生成
  namedWindow(win_src, WINDOW_AUTOSIZE);
  // namedWindow(win_dst, WINDOW_AUTOSIZE);

  vector<vector<Point>> contours;
  findContours(img_src, contours, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);
  double area = contourArea(Mat(contours[0])); // 面積
  double perimeter = arcLength(Mat(contours[0]), true); // 周囲長
  double roundness = 4 * M_PI * area / perimeter / perimeter; // 円形度
  cout << area << " " << perimeter << " " << roundness << endl;

  imshow(win_src, img_src); // 入力画像を表示
  // imshow(win_dst, img_dst); // 出力画像を表示
  // imwrite(file_dst, img_dst); // 処理結果の保存

  waitKey(0); // キー入力待ち
  return 0;
}
